lib/charms/hpc_libs/v0/slurm_ops.py

```diff
@@ -197,9 +180,244 @@ def config_name(self) -> str:
+
+    def __init__(self, ops_manager: SlurmOpsManager) -> None:
+        self.service = ops_manager.service_manager_for(ServiceType.PROMETHEUS_EXPORTER)
+
+
+class SlurmManagerBase:
+    """Base manager for Slurm services."""
+
+    def __init__(self, service: ServiceType, snap: bool = False) -> None:
+        if not snap:
+            raise SlurmOpsError("deb packaging is currently unimplemented")
+        self._ops_manager = SnapManager()
+        self.service = self._ops_manager.service_manager_for(service)
+        self.munge = MungeManager(self._ops_manager)
+        self.exporter = PrometheusExporterManager(self._ops_manager)
+        self.install = self._ops_manager.install
+        self.version = self._ops_manager.version
+
+    @property
+    def hostname(self) -> str:
+        """The hostname where this manager is running."""
+        return socket.gethostname().split(".")[0]
+
+
+class SlurmctldManager(SlurmManagerBase):
+    """Manager for the Slurmctld service."""
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(service=ServiceType.SLURMCTLD, *args, **kwargs)
+        self._config_path = self._ops_manager.slurm_path / "slurm.conf"
+
+    @contextmanager
+    def config(self) -> slurmconfig.SlurmConfig:
+        """Get the config manager of slurmctld."""
+        with slurmconfig.edit(self._config_path) as config:
+            yield config
+
+
+class SlurmdManager(SlurmManagerBase):
+    """Manager for the Slurmd service."""
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(service=ServiceType.SLURMD, *args, **kwargs)
+        self._env_manager = self._ops_manager._env_manager_for(ServiceType.SLURMD)
+
+    @property
+    def config_server(self) -> str:
+        """Get the config server address of this Slurmd node."""
+        return self._env_manager.get("config-server")
+
+    @config_server.setter
+    def config_server(self, addr: str) -> None:
+        """Set the config server address of this Slurmd node."""
+        self._env_manager.set({"config-server": addr})
+
+    @config_server.deleter
+    def config_server(self) -> None:
+        """Unset the config server address of this Slurmd node."""
+        self._env_manager.unset("config-server")
+
+
+class SlurmdbdManager(SlurmManagerBase):
+    """Manager for the Slurmdbd service."""
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(service=ServiceType.SLURMDBD, *args, **kwargs)
+        self._config_path = self._ops_manager.slurm_path / "slurmdbd.conf"
+
+    @contextmanager
+    def config(self) -> slurmdbdconfig.SlurmdbdConfig:
+        """Get the config manager of slurmctld."""
+        with slurmdbdconfig.edit(self._config_path) as config:
+            yield config
+
+
+class SlurmrestdManager(SlurmManagerBase):
+    """Manager for the Slurmrestd service."""
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(service=ServiceType.SLURMRESTD, *args, **kwargs)
```

Comment:
Should we instead have these defined in their respective charms instead? 

I know we discussed earlier about having these custom methods for managing configuration of the services, but what I'm concerned about is now that we have these defined here, anytime we update one manager, we end up needing to update the lib across all the charms.

This way we can also discuss the implementation of each manager in its own PR against the relevant charm.

Reply:
I like the current design because it abstracts away the interactions with `OpsManager`, and hides everything behind a `snap: bool` flag. Moving this to the charms would do the opposite, forcing us to use `OpsManager` from the charms.

Reply:
Okay, sounds good to me then. Having `config` exposed as a context manager, we can just update the the current configuration using the `update(...)` method. 

We'll still want to inherit from the dedicated service managers in their respective charms, correct? We'll still need to compose certain functionality on `SlurmctldManager` like the ability to edit the cgroup or OCI image configuration. Are you envisioning that this functionality would be defined within `hpc_libs`, or would you see this as something we add within the slurmctld charm?

Reply:
I'm envisioning that any additional functionality would start on a "testing phase" in the deb version of the charm, and if we determine that the functionality is good enough and also useful for the snap version, we would move it to `hpc_libs`